File: apps/mobile/lib/services/plate_settings.dart
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../utils/plates.dart';

/// Device-persistent settings for the plate calculator: the bar weight and the
/// plates the user owns. Loaded once at startup and read synchronously from the
/// workout screen. A [ValueListenable] so settings UIs rebuild on change.
///
/// Stored locally (SharedPreferences) like the auth token — the backend data
/// layer isn't user-scoped yet, so this is per-device.
class PlateSettings extends ChangeNotifier {
  PlateSettings._();
  static final PlateSettings instance = PlateSettings._();

  static const _barKey = 'plate_bar_weight';
  static const _platesKey = 'plate_available';

  double _barWeight = defaultBarWeight;
  List<double> _availablePlates = List.of(defaultPlates);

  double get barWeight => _barWeight;
  List<double> get availablePlates => List.unmodifiable(_availablePlates);

  /// Load persisted values. Never throws — falls back to defaults.
  Future<void> load() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final bar = prefs.getDouble(_barKey);
      final plates = prefs.getStringList(_platesKey);
      if (bar != null && bar > 0) _barWeight = bar;
      if (plates != null && plates.isNotEmpty) {
        final parsed = plates
            .map(double.tryParse)
            .whereType<double>()
            .where((p) => p > 0)
            .toList()
          ..sort((a, b) => b.compareTo(a));
        if (parsed.isNotEmpty) _availablePlates = parsed;
      }
    } catch (_) {
      // Storage unavailable — keep defaults.
    }
    notifyListeners();
  }

  /// Update + persist. Best-effort persistence; in-memory always succeeds.
  Future<void> update({double? barWeight, List<double>? availablePlates}) async {
    if (barWeight != null && barWeight > 0) _barWeight = barWeight;
    if (availablePlates != null) {
      final cleaned = availablePlates.where((p) => p > 0).toSet().toList()
        ..sort((a, b) => b.compareTo(a));
      if (cleaned.isNotEmpty) _availablePlates = cleaned;
    }
    notifyListeners();
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setDouble(_barKey, _barWeight);
      await prefs.setStringList(
        _platesKey,
        _availablePlates.map((p) => p.toString()).toList(),
      );
    } catch (_) {
      // Couldn't persist; values still live for this session.
    }
  }

  /// Reset to the spec defaults.
  Future<void> resetToDefaults() =>
      update(barWeight: defaultBarWeight, availablePlates: List.of(defaultPlates));
}

File: apps/native-mobile/src/lib/plateSettings.ts
// Device-persistent plate-calculator settings (bar weight + owned plates).
// Ported from services/plate_settings.dart — a singleton store with listeners
// (subscribe via usePlateSettings) backed by AsyncStorage.
import AsyncStorage from "@react-native-async-storage/async-storage";
import { defaultBarWeight, defaultPlates } from "../utils/plates";

const BAR_KEY = "plate_bar_weight";
const PLATES_KEY = "plate_available";

class PlateSettingsStore {
  private _barWeight = defaultBarWeight;
  private _availablePlates: number[] = [...defaultPlates];
  private listeners = new Set<() => void>();

  get barWeight(): number {
    return this._barWeight;
  }
  get availablePlates(): number[] {
    return [...this._availablePlates];
  }

  subscribe(fn: () => void): () => void {
    this.listeners.add(fn);
    return () => this.listeners.delete(fn);
  }
  private emit(): void {
    for (const fn of this.listeners) fn();
  }

  /** Load persisted values. Never throws — falls back to defaults. */
  async load(): Promise<void> {
    try {
      const [barRaw, platesRaw] = await Promise.all([
        AsyncStorage.getItem(BAR_KEY),
        AsyncStorage.getItem(PLATES_KEY),
      ]);
      const bar = barRaw != null ? Number(barRaw) : null;
      if (bar != null && bar > 0) this._barWeight = bar;
      if (platesRaw != null) {
        const parsed = (JSON.parse(platesRaw) as unknown[])
          .map((p) => Number(p))
          .filter((p) => Number.isFinite(p) && p > 0)
          .sort((a, b) => b - a);
        if (parsed.length > 0) this._availablePlates = parsed;
      }
    } catch {
      // keep defaults
    }
    this.emit();
  }

  /** Update + persist. Best-effort persistence; in-memory always succeeds. */
  async update(opts: { barWeight?: number; availablePlates?: number[] }): Promise<void> {
    if (opts.barWeight != null && opts.barWeight > 0) this._barWeight = opts.barWeight;
    if (opts.availablePlates != null) {
      const cleaned = Array.from(new Set(opts.availablePlates.filter((p) => p > 0))).sort((a, b) => b - a);
      if (cleaned.length > 0) this._availablePlates = cleaned;
    }
    this.emit();
    try {
      await AsyncStorage.setItem(BAR_KEY, String(this._barWeight));
      await AsyncStorage.setItem(PLATES_KEY, JSON.stringify(this._availablePlates));
    } catch {
      // values still live for this session
    }
  }

  resetToDefaults(): Promise<void> {
    return this.update({ barWeight: defaultBarWeight, availablePlates: [...defaultPlates] });
  }
}

export const plateSettings = new PlateSettingsStore();
